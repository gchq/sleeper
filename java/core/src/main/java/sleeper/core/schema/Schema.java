/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.schema;

import sleeper.core.schema.type.PrimitiveType;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.TreeSet;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * A Schema describes the fields found in a particular table in a Sleeper instance.
 */
public class Schema {
    private final List<Field> rowKeyFields;
    private final List<Field> sortKeyFields;
    private final List<Field> valueFields;

    private Schema(Builder builder) {
        rowKeyFields = validateRowKeys(builder.rowKeyFields);
        sortKeyFields = validateSortKeys(builder.sortKeyFields);
        valueFields = validateValueKeys(builder.valueFields);
        validateNoDuplicates(streamAllFields(rowKeyFields, sortKeyFields, valueFields));
    }

    public static Builder builder() {
        return new Builder();
    }

    /**
     * Creates an instance of this class using a schema file.
     *
     * @param  schemaPath  the path to the schema file
     * @return             an instance of this class
     * @throws IOException if an error occurs reading from the file
     */
    public static Schema load(Path schemaPath) throws IOException {
        return loadFromString(Files.readString(schemaPath));
    }

    /**
     * Creates an instance of this class using a JSON string.
     *
     * @param  schemaJson the JSON string representing a schema
     * @return            an instance of this class
     */
    public static Schema loadFromString(String schemaJson) {
        return new SchemaSerDe().fromJson(schemaJson);
    }

    public List<Field> getRowKeyFields() {
        return rowKeyFields;
    }

    public List<Field> getSortKeyFields() {
        return sortKeyFields;
    }

    public List<Field> getValueFields() {
        return valueFields;
    }

    public List<PrimitiveType> getRowKeyTypes() {
        return getMappedFields(rowKeyFields, f -> (PrimitiveType) f.getType());
    }

    public List<PrimitiveType> getSortKeyTypes() {
        return getMappedFields(sortKeyFields, f -> (PrimitiveType) f.getType());
    }

    public List<String> getRowKeyFieldNames() {
        return getMappedFields(rowKeyFields, Field::getName);
    }

    public List<String> getSortKeyFieldNames() {
        return getMappedFields(sortKeyFields, Field::getName);
    }

    public List<String> getValueFieldNames() {
        return getMappedFields(valueFields, Field::getName);
    }

    public List<String> getAllFieldNames() {
        return getMappedFields(streamAllFields(), Field::getName);
    }

    private <T> List<T> getMappedFields(List<Field> fields, Function<Field, T> mapping) {
        return getMappedFields(fields.stream(), mapping);
    }

    private <T> List<T> getMappedFields(Stream<Field> fields, Function<Field, T> mapping) {
        return fields
                .map(mapping)
                .collect(Collectors.toUnmodifiableList());
    }

    public List<Field> getAllFields() {
        return streamAllFields().collect(Collectors.toUnmodifiableList());
    }

    /**
     * Streams all fields in the schema. This includes row key, sort key, and value fields.
     *
     * @return a stream of all fields
     */
    public Stream<Field> streamAllFields() {
        return streamAllFields(rowKeyFields, sortKeyFields, valueFields);
    }

    /**
     * Gets a field by name.
     *
     * @param  fieldName the name of the field
     * @return           a field, or an empty optional if the field does not exist
     */
    public Optional<Field> getField(String fieldName) {
        return streamAllFields()
                .filter(f -> f.getName().equals(fieldName))
                .findFirst();
    }

    /**
     * Saves this schema to a file.
     *
     * @param  path        the path of the file
     * @throws IOException if an error occured when writing the file
     */
    public void save(Path path) throws IOException {
        Files.writeString(path, new SchemaSerDe().toJson(this));
    }

    @Override
    public String toString() {
        return "Schema{" + "rowKeyFields=" + rowKeyFields + ", sortKeyFields=" + sortKeyFields + ", valueFields=" + valueFields + '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Schema schema = (Schema) o;

        return Objects.equals(rowKeyFields, schema.rowKeyFields) &&
                Objects.equals(sortKeyFields, schema.sortKeyFields) &&
                Objects.equals(valueFields, schema.valueFields);
    }

    @Override
    public int hashCode() {
        return Objects.hash(rowKeyFields, sortKeyFields, valueFields);
    }

    /**
     * Builder to create a schema object.
     */
    public static final class Builder {
        private List<Field> rowKeyFields;
        private List<Field> sortKeyFields;
        private List<Field> valueFields;

        private Builder() {
        }

        /**
         * Sets the row key fields.
         *
         * @param  rowKeyFields the row key fields
         * @return              the builder
         */
        public Builder rowKeyFields(List<Field> rowKeyFields) {
            this.rowKeyFields = rowKeyFields;
            return this;
        }

        /**
         * Sets the row key fields.
         *
         * @param  rowKeyFields the row key fields
         * @return              the builder
         */
        public Builder rowKeyFields(Field... rowKeyFields) {
            return rowKeyFields(Arrays.asList(rowKeyFields));
        }

        /**
         * Sets the sort key fields.
         *
         * @param  sortKeyFields the sort key fields
         * @return               the builder
         */
        public Builder sortKeyFields(List<Field> sortKeyFields) {
            this.sortKeyFields = sortKeyFields;
            return this;
        }

        /**
         * Sets the sort key fields.
         *
         * @param  sortKeyFields the sort key fields
         * @return               the builder
         */
        public Builder sortKeyFields(Field... sortKeyFields) {
            return sortKeyFields(Arrays.asList(sortKeyFields));
        }

        /**
         * Sets the value fields.
         *
         * @param  valueFields the value fields
         * @return             the builder
         */
        public Builder valueFields(List<Field> valueFields) {
            this.valueFields = valueFields;
            return this;
        }

        /**
         * Sets the value fields.
         *
         * @param  valueFields the value fields
         * @return             the builder
         */
        public Builder valueFields(Field... valueFields) {
            return valueFields(Arrays.asList(valueFields));
        }

        public Schema build() {
            return new Schema(this);
        }
    }

    private static List<Field> validateRowKeys(List<Field> fields) {
        if (fields == null || fields.isEmpty()) {
            throw new IllegalArgumentException("Must have at least one row key field");
        }
        if (fields.stream().anyMatch(field -> !(field.getType() instanceof PrimitiveType))) {
            throw new IllegalArgumentException("Row key fields must have a primitive type");
        }
        return makeImmutable(fields);
    }

    private static List<Field> validateSortKeys(List<Field> fields) {
        if (fields == null) {
            return Collections.emptyList();
        }
        if (fields.stream().anyMatch(field -> !(field.getType() instanceof PrimitiveType))) {
            throw new IllegalArgumentException("Sort key fields must have a primitive type");
        }
        return makeImmutable(fields);
    }

    private static List<Field> validateValueKeys(List<Field> fields) {
        if (fields == null) {
            return Collections.emptyList();
        }
        return makeImmutable(fields);
    }

    private static List<Field> makeImmutable(List<Field> fields) {
        return Collections.unmodifiableList(new ArrayList<>(fields));
    }

    private static Stream<Field> streamAllFields(
            List<Field> rowKeyFields, List<Field> sortKeyFields, List<Field> valueFields) {
        return Stream.of(rowKeyFields, sortKeyFields, valueFields)
                .flatMap(List::stream);
    }

    private static void validateNoDuplicates(Stream<Field> fields) {
        Set<String> foundNames = new HashSet<>();
        Set<String> duplicates = new TreeSet<>();
        fields.forEach(field -> {
            boolean isNew = foundNames.add(field.getName());
            if (!isNew) {
                duplicates.add(field.getName());
            }
        });
        if (!duplicates.isEmpty()) {
            throw new IllegalArgumentException("Found duplicate field names: " + duplicates);
        }
    }
}
