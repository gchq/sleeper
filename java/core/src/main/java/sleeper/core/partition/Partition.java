/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.partition;

import sleeper.core.key.Key;
import sleeper.core.range.Region;
import sleeper.core.range.RegionCanonicaliser;
import sleeper.core.schema.Schema;

import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

/**
 * A Partition is a Region of key space, with additional information that allows
 * the position of the partition in the tree of partitions to be identified,
 * e.g. it has a unique id, the id of its parent partition, the id of any child
 * partitions, and if it has been split the dimension it was split on.
 * <p>
 * The {@link Region} must be in canonical form, i.e. all the ranges must contain
 * their minimum but not contain their maximum.
 */
public class Partition {
    private final Region region;
    private final String id;
    private final boolean leafPartition;
    private final String parentPartitionId;
    private final List<String> childPartitionIds;
    private final int dimension;

    private Partition(Partition.Builder builder) {
        region = Objects.requireNonNull(builder.region, "region must not be null");
        id = Objects.requireNonNull(builder.id, "id must not be null");
        leafPartition = builder.leafPartition;
        parentPartitionId = builder.parentPartitionId;
        childPartitionIds = Optional.ofNullable(builder.childPartitionIds).orElse(Collections.emptyList());
        dimension = builder.dimension;
        if (!RegionCanonicaliser.isRegionInCanonicalForm(region)) {
            throw new IllegalArgumentException("Region must be in canonical form");
        }
    }

    public static Builder builder() {
        return new Builder();
    }

    /**
     * Checks if a row key is in this partition. This depends on the ranges the partition covers.
     *
     * @param  schema schema of the Sleeper table
     * @param  rowKey values of row key to check
     * @return        true if the row key is in the partition
     */
    public boolean isRowKeyInPartition(Schema schema, Key rowKey) {
        return region.isKeyInRegion(schema, rowKey);
    }

    /**
     * Checks if a region overlaps this partition.
     *
     * @param  otherRegion region to check
     * @return             true if any part of the region is in this partition
     */
    public boolean doesRegionOverlapPartition(Region otherRegion) {
        return region.doesRegionOverlap(otherRegion);
    }

    public Region getRegion() {
        return region;
    }

    public String getId() {
        return id;
    }

    public boolean isLeafPartition() {
        return leafPartition;
    }

    public String getParentPartitionId() {
        return parentPartitionId;
    }

    public List<String> getChildPartitionIds() {
        return childPartitionIds;
    }

    public int getDimension() {
        return dimension;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Partition partition = (Partition) o;

        return leafPartition == partition.leafPartition
                && Objects.equals(region, partition.region)
                && Objects.equals(id, partition.id)
                && Objects.equals(parentPartitionId, partition.parentPartitionId)
                && Objects.equals(childPartitionIds, partition.childPartitionIds)
                && dimension == partition.getDimension();
    }

    @Override
    public int hashCode() {
        return Objects.hash(region, id, leafPartition,
                parentPartitionId, childPartitionIds, dimension);
    }

    @Override
    public String toString() {
        return "Partition{"
                + "region=" + region
                + ", id='" + id + '\''
                + ", leafPartition=" + leafPartition
                + ", parentPartitionId='" + parentPartitionId + '\''
                + ", childPartitionIds=" + childPartitionIds
                + ", dimension=" + dimension
                + '}';
    }

    public Builder toBuilder() {
        return builder()
                .region(region)
                .id(id)
                .leafPartition(leafPartition)
                .childPartitionIds(childPartitionIds)
                .parentPartitionId(parentPartitionId)
                .dimension(dimension);
    }

    /**
     * Builder to create a partition object.
     */
    public static final class Builder {
        private Region region;
        private String id;
        private boolean leafPartition;
        private String parentPartitionId;
        private List<String> childPartitionIds;
        private int dimension = -1; // -1 used to indicate that it has not been split yet; when it has been split, indicates which dimension was used to split on.

        private Builder() {
        }

        /**
         * Sets the region covered by the partition.
         *
         * @param  region the region
         * @return        the builder
         */
        public Builder region(Region region) {
            this.region = region;
            return this;
        }

        /**
         * Sets the ID of the partition.
         *
         * @param  id a unique identifier
         * @return    the builder
         */
        public Builder id(String id) {
            this.id = id;
            return this;
        }

        /**
         * Sets whether this is a leaf partition.
         *
         * @param  leafPartition true if the partition has no child partitions, false otherwise
         * @return               the builder
         */
        public Builder leafPartition(boolean leafPartition) {
            this.leafPartition = leafPartition;
            return this;
        }

        /**
         * Sets the parent of the partition. This is the ID of the partition that was split to create this partition.
         * Can be null if this is the root partition of the Sleeper table.
         *
         * @param  parentPartitionId the ID of the parent partition
         * @return                   the builder
         */
        public Builder parentPartitionId(String parentPartitionId) {
            this.parentPartitionId = parentPartitionId;
            return this;
        }

        /**
         * Sets the children of the partition. This is the IDs of any partitions that were created by splitting this
         * partition. This will default to an empty list, and must be empty for a leaf partition.
         *
         * @param  childPartitionIds the IDs of the child partitions
         * @return                   the builder
         */
        public Builder childPartitionIds(List<String> childPartitionIds) {
            this.childPartitionIds = childPartitionIds;
            return this;
        }

        /**
         * Sets the dimension this partition was split on. If this partition has been split, this should be the index of
         * the row key in the schema that this partition was split on.
         *
         * @param  dimension the index of the row key used as the split point
         * @return           the builder
         */
        public Builder dimension(int dimension) {
            this.dimension = dimension;
            return this;
        }

        public Partition build() {
            return new Partition(this);
        }

        public String getId() {
            return id;
        }

        public Region getRegion() {
            return region;
        }
    }
}
