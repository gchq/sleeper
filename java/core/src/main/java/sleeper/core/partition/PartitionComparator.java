/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.partition;

import sleeper.core.schema.Schema;

import java.util.Arrays;
import java.util.Comparator;

/**
 * Comparator for organised paritions in correct orientation.
 */
public class PartitionComparator implements Comparator<Partition> {

    private Schema schema;

    PartitionComparator(Schema schema) {
        this.schema = schema;
    }

    @Override
    public int compare(Partition partition1, Partition partition2) {
        String rowKey = schema.getRowKeyFields().get(0).getName();
        Object minA = partition1.getRegion().getRange(rowKey).getMin();
        Object minB = partition2.getRegion().getRange(rowKey).getMin();

        if (minA instanceof Long && minB instanceof Long) {
            return ((Long) minA).compareTo((Long) minB);
        }

        if (minA instanceof Integer && minB instanceof Integer) {
            return ((Integer) minA).compareTo((Integer) minB);
        }

        if (minA instanceof String && minB instanceof String) {
            return ((String) minA).compareTo((String) minB);
        }

        if (minA instanceof byte[] && minB instanceof byte[]) {
            return Arrays.compare((byte[]) minA, (byte[]) minB);
        }

        throw new PartitionTreeException("Unable to compare partitions due to range elements being an " +
                "unexpected class type. minA: " + minA.getClass() + ", minB: " + minB.getClass());
    }
}
