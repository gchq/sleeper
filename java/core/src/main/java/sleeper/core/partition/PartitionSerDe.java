/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.partition;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonArray;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;
import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;

import sleeper.core.range.Region;
import sleeper.core.range.RegionSerDe.RegionJsonSerDe;
import sleeper.core.schema.Schema;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

/**
 * Serialises a partition to and from a JSON string.
 */
public class PartitionSerDe {
    public static final String PARTITION_ID = "partitionId";
    public static final String IS_LEAF_PARTITION = "isLeafPartition";
    public static final String PARENT_PARTITION_ID = "parentPartitionId";
    public static final String CHILD_PARTITION_IDS = "childPartitionIds";
    public static final String REGION = "region";
    public static final String DIMENSION = "dimension";

    private final Gson gson;
    private final Gson gsonPrettyPrinting;

    public PartitionSerDe(Schema schema) {
        GsonBuilder builder = new GsonBuilder()
                .registerTypeAdapter(Partition.class, new PartitionJsonSerDe(schema))
                .serializeNulls();
        this.gson = builder.create();
        this.gsonPrettyPrinting = builder.setPrettyPrinting().create();
    }

    /**
     * Serialises a partition to JSON.
     *
     * @param  partition the partition
     * @return           a JSON representation of the partition
     */
    public String toJson(Partition partition) {
        return gson.toJson(partition);
    }

    /**
     * Serialises a partition to JSON.
     *
     * @param  partition   the partition
     * @param  prettyPrint true if the JSON should be formatted for readability
     * @return             a JSON representation of the partition
     */
    public String toJson(Partition partition, boolean prettyPrint) {
        if (prettyPrint) {
            return gsonPrettyPrinting.toJson(partition);
        }
        return toJson(partition);
    }

    /**
     * Deserialises a partition from JSON.
     *
     * @param  jsonSchema the JSON
     * @return            the partition represented by the JSON
     */
    public Partition fromJson(String jsonSchema) {
        return gson.fromJson(jsonSchema, Partition.class);
    }

    /**
     * A GSON plugin to serialise/deserialise a partition.
     */
    public static class PartitionJsonSerDe implements JsonSerializer<Partition>, JsonDeserializer<Partition> {
        private final RegionJsonSerDe regionJsonSerDe;

        public PartitionJsonSerDe(Schema schema) {
            this.regionJsonSerDe = new RegionJsonSerDe(schema);
        }

        @Override
        @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE")
        public JsonElement serialize(Partition partition, java.lang.reflect.Type typeOfSrc, JsonSerializationContext context) {
            JsonObject json = new JsonObject();
            json.addProperty(PARTITION_ID, partition.getId());
            json.addProperty(IS_LEAF_PARTITION, partition.isLeafPartition());
            json.addProperty(PARENT_PARTITION_ID, partition.getParentPartitionId());
            JsonArray childPartitionIds = new JsonArray();
            if (null != partition.getChildPartitionIds()) {
                for (String childId : partition.getChildPartitionIds()) {
                    childPartitionIds.add(childId);
                }
            }
            json.add(CHILD_PARTITION_IDS, childPartitionIds);
            json.add(REGION, regionJsonSerDe.serialize(partition.getRegion(), null, context));
            json.addProperty(DIMENSION, partition.getDimension());
            return json;
        }

        @Override
        public Partition deserialize(JsonElement jsonElement, java.lang.reflect.Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
            if (!jsonElement.isJsonObject()) {
                throw new JsonParseException("Expected JsonObject, got " + jsonElement);
            }
            JsonObject json = jsonElement.getAsJsonObject();
            String partitionId = json.get(PARTITION_ID).getAsString();
            boolean isLeafPartition = json.get(IS_LEAF_PARTITION).getAsBoolean();
            String parentPartitionId = null;
            if (json.has(PARENT_PARTITION_ID)) {
                if (!json.get(PARENT_PARTITION_ID).isJsonNull()) {
                    parentPartitionId = json.get(PARENT_PARTITION_ID).getAsString();

                }
            }
            JsonArray childPartitionIdsArray = json.get(CHILD_PARTITION_IDS).getAsJsonArray();
            List<String> childPartitionIds = new ArrayList<>();
            Iterator<JsonElement> it = childPartitionIdsArray.iterator();
            while (it.hasNext()) {
                JsonElement element = it.next();
                childPartitionIds.add(element.getAsString());
            }
            Region region = regionJsonSerDe.deserialize(json.get(REGION), null, context);
            int dimension = json.get(DIMENSION).getAsInt();
            return Partition.builder()
                    .region(region)
                    .id(partitionId)
                    .leafPartition(isLeafPartition)
                    .parentPartitionId(parentPartitionId)
                    .childPartitionIds(childPartitionIds)
                    .dimension(dimension)
                    .build();
        }
    }
}
