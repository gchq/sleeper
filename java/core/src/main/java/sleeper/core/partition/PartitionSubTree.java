/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.partition;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Objects;

/**
 * Generates a subtree of PartitionTree for given number of leaf nodes.
 */
public class PartitionSubTree extends PartitionTree {
    private final ArrayList<String> resetLeafIds = new ArrayList<>();

    public PartitionSubTree(PartitionTree originalTree, int leafPartitionCount) {
        super(List.of(originalTree.getRootPartition()));

        // Check loop has count incremented by 1 to account for root partition that always must exist
        while (idToPartition.values().size() < leafPartitionCount + 1) {
            resetLeafIds.clear();
            Collection<Partition> presentBatch = List.copyOf(idToPartition.values());
            presentBatch.forEach(partition -> {
                partition.getChildPartitionIds().forEach(
                        partitionId -> {
                            if (!idToPartition.containsKey(partitionId)) {
                                idToPartition.put(partitionId, originalTree.getPartition(partitionId));
                                resetLeafIds.add(partitionId);
                            }
                        });
            });
        }

        // Last gathered selection of ids are set to be leaves of new tree.
        resetLeafIds.forEach(leafId -> {
            idToPartition.put(leafId, idToPartition.get(leafId)
                    .toBuilder()
                    .leafPartition(true)
                    .build());
        });
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        PartitionSubTree that = (PartitionSubTree) o;
        return idToPartition.equals(that.idToPartition);
    }

    @Override
    public int hashCode() {
        return Objects.hash(idToPartition);
    }
}
