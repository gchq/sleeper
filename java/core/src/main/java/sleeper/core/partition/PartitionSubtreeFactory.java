/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.partition;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

/**
 * Utility class to generate subtree from given leaf partition requirements.
 */
public class PartitionSubtreeFactory {

    private PartitionSubtreeFactory() {
    }

    /**
     * Generates a subtree from a given tree with a target leaf count. Actual result may exceed the count requested
     * as presently will go to a depth until the count is matched or exceeded.
     *
     * @param  originalTree       source from which the new sub tree is to be created
     * @param  leafPartitionCount amount of leaves to be contained in the new tree at a minimum
     * @return                    newly generated sub tree
     */
    public static PartitionTree createSubtree(PartitionTree originalTree, int leafPartitionCount) throws PartitionTreeException {
        if (leafPartitionCount > originalTree.getLeafPartitions().size()) {
            throw new PartitionTreeException("Requested size of " + leafPartitionCount + " is greater than input tree capacity");
        }
        ArrayList<String> resetLeafIds = new ArrayList<>();
        PartitionTree subtree = new PartitionTree(List.of(originalTree.getRootPartition()));
        resetLeafIds.add(originalTree.getRootPartition().getId());

        // Check loop has count incremented by 1 to account for root partition that always must exist
        while (subtree.idToPartition.values().size() < leafPartitionCount + 1) {
            resetLeafIds.clear();
            Collection<Partition> presentBatch = List.copyOf(subtree.idToPartition.values());
            presentBatch.forEach(partition -> {
                partition.getChildPartitionIds().forEach(
                        partitionId -> {
                            if (!subtree.idToPartition.containsKey(partitionId)) {
                                subtree.idToPartition.put(partitionId, originalTree.getPartition(partitionId));
                                resetLeafIds.add(partitionId);
                            }
                        });
            });
        }

        // Last gathered selection of ids have their key details reset to act as leaves of new tree.
        resetLeafIds.forEach(leafId -> {
            subtree.idToPartition.put(leafId, adjustToLeafStatus(subtree.idToPartition.get(leafId)));

        });

        // If root is the only partition, return a new sub tree with root corrected for no leaves
        if (subtree.getAllPartitions().size() == 1) {
            return new PartitionTree(subtree.getAllPartitions());
        }

        return subtree;
    }

    private static Partition adjustToLeafStatus(Partition partitionIn) {
        return partitionIn.toBuilder()
                .leafPartition(true)
                .childPartitionIds(Collections.emptyList())
                .dimension(-1)
                .build();
    }
}
