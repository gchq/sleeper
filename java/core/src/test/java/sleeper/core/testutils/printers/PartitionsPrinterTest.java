/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.core.testutils.printers;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import sleeper.core.partition.PartitionsBuilder;
import sleeper.core.schema.Schema;
import sleeper.core.schema.type.LongType;

import java.util.Map;

import static sleeper.core.schema.SchemaTestHelper.schemaWithKey;

public class PartitionsPrinterTest {

    private final Schema schema = schemaWithKey("key", new LongType());
    private final PartitionsBuilder partitions = new PartitionsBuilder(schema);

    @Test
    void shouldPrintTreeWith8LevelsOfSplits() {
        // Given
        partitions.rootFirst("root")
                .splitToNewChildren("root", "L", "R", 50L)
                .splitToNewChildren("L", "LL", "LR", 25L)
                .splitToNewChildren("R", "RL", "RR", 75L)
                .splitToNewChildren("LL", "LLL", "LLR", 12L)
                .splitToNewChildren("LR", "LRL", "LRR", 37L)
                .splitToNewChildren("RL", "RLL", "RLR", 62L)
                .splitToNewChildren("RR", "RRL", "RRR", 87L);

        // When
        String printed = PartitionsPrinter.printPartitions(schema, partitions.buildTree());

        // Then see approved output
        Approvals.verify(printed);
    }

    @Test
    void shouldRenamePartitionsByLocation() {
        // Given
        partitions.rootFirst("base")
                .splitToNewChildren("base", "l", "r", 50L)
                .splitToNewChildren("l", "ll", "lr", 25L)
                .splitToNewChildren("r", "rl", "rr", 75L)
                .splitToNewChildren("ll", "1", "2", 12L)
                .splitToNewChildren("lr", "3", "4", 37L)
                .splitToNewChildren("rl", "5", "6", 62L)
                .splitToNewChildren("rr", "7", "8", 87L);

        // When
        String printed = PartitionsPrinter.printPartitions(schema, partitions.buildTree());

        // Then see approved output
        Approvals.verify(printed);
    }

    @Test
    void shouldPrintDifferentPartitionsForOneTable() {
        // Given
        PartitionsBuilder partitions1 = new PartitionsBuilder(schema).rootFirst("A")
                .splitToNewChildren("A", "B", "C", 10L);
        PartitionsBuilder partitions2 = new PartitionsBuilder(schema).rootFirst("1")
                .splitToNewChildren("1", "2", "3", 20L);

        // When
        String printed = PartitionsPrinter.printTablePartitionsExpectingIdentical(schema, Map.of(
                "table-1", partitions1.buildTree(),
                "table-2", partitions2.buildTree(),
                "table-3", partitions1.buildTree()));

        // Then see approved output
        Approvals.verify(printed);
    }
}
