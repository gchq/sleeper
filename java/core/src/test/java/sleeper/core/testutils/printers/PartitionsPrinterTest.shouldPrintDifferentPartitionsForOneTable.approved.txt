Different for one table
Leaf partition at L:
{"key":{"min":-9223372036854775808,"minInclusive":true,"max":20,"maxInclusive":false},"stringsBase64Encoded":true}
Leaf partition at R:
{"key":{"min":20,"minInclusive":true,"max":null,"maxInclusive":false},"stringsBase64Encoded":true}
Partition at root:
{"key":{"min":-9223372036854775808,"minInclusive":true,"max":null,"maxInclusive":false},"stringsBase64Encoded":true}

Same for 2 tables
Leaf partition at L:
{"key":{"min":-9223372036854775808,"minInclusive":true,"max":10,"maxInclusive":false},"stringsBase64Encoded":true}
Leaf partition at R:
{"key":{"min":10,"minInclusive":true,"max":null,"maxInclusive":false},"stringsBase64Encoded":true}
Partition at root:
{"key":{"min":-9223372036854775808,"minInclusive":true,"max":null,"maxInclusive":false},"stringsBase64Encoded":true}

